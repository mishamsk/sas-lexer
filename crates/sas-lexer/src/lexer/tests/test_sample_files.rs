use super::{
    super::{lex_program, LexResult},
    util::{error_to_string, token_to_string},
};
use insta::assert_yaml_snapshot;
use rstest::rstest;
use std::{fs, path::PathBuf};

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

#[rstest]
fn test_snapshots(#[files("src/lexer/tests/samples/**/*.sas")] path: PathBuf) {
    // Compute the absolute path of the prefix

    let prefix = fs::canonicalize(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/src/lexer/tests/samples/"
    ))
    .unwrap();

    let snap_name = path.strip_prefix(&prefix).unwrap();
    let snap_name_str = snap_name
        .to_str()
        .unwrap()
        .replace(std::path::MAIN_SEPARATOR, "__");

    set_snapshot_suffix!("{}", snap_name_str);

    let contents = fs::read_to_string(&path).unwrap();
    let LexResult {
        buffer: tok_buffer,
        errors,
        ..
    } = lex_program(&contents).unwrap();
    let tokens: Vec<String> = tok_buffer
        .into_iter()
        .map(|tidx| token_to_string(tidx, &tok_buffer, &contents))
        .collect();

    assert_yaml_snapshot!("tokens", tokens);

    if errors.is_empty() {
        insta::allow_duplicates! {
            assert_yaml_snapshot!("errors_snapshot", @r###"
            errors_snapshot
            "###);
        }
    } else {
        let error_strings = errors
            .iter()
            .map(|error| error_to_string(error, &tok_buffer, &contents).to_string())
            .collect::<Vec<String>>();
        assert_yaml_snapshot!("errors_snapshot", error_strings);
    }
}

/// Tests that tokens cover all offsets in the file from 0 to the end
#[rstest]
fn test_full_coverage(#[files("src/lexer/tests/samples/**/*.sas")] path: PathBuf) {
    let contents = fs::read_to_string(&path).unwrap();

    let LexResult {
        buffer: tok_buffer, ..
    } = lex_program(&contents).unwrap();

    let mut end = 0;

    for token in &tok_buffer {
        // Check that the token starts where the previous token ended
        assert_eq!(
            tok_buffer
                .get_token_start(token)
                .expect("wrong token")
                .get(),
            end,
            "Token <{}> does not start where the previous token ended",
            token_to_string(token, &tok_buffer, &contents)
        );

        // Set the new end
        end = tok_buffer.get_token_end(token).expect("wrong token").get();

        // Check that the end is greater than the start
        assert!(
            end >= tok_buffer
                .get_token_start(token)
                .expect("wrong token")
                .get(),
            "Token <{}> has an end before the start",
            token_to_string(token, &tok_buffer, &contents)
        );
    }

    // Check that the last token ends at the end of the file
    assert_eq!(end, contents.len() as u32);
}
