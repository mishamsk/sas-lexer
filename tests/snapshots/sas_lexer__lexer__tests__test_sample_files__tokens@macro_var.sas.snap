---
source: crates/sas-lexer/src/lexer/tests/test_sample_files.rs
expression: tokens
---
- "[@0,0:1=\"&\",<AMP>,L1:C0-L1:C1,chl=<D>,pl=<None>]"
- "[@1,1:2=\"\n\",<WS>,L1:C1-L1:C2,chl=<H>,pl=<None>]"
- "[@2,2:6=\"&&&&\",<AMP>,L2:C0-L2:C4,chl=<D>,pl=<None>]"
- "[@3,6:7=\"\n\",<WS>,L2:C4-L2:C5,chl=<H>,pl=<None>]"
- "[@4,7:8=\"&\",<AMP>,L3:C0-L3:C1,chl=<D>,pl=<None>]"
- "[@5,8:9=\"9\",<IntegerLiteral>,L3:C1-L3:C2,chl=<D>,pl=9]"
- "[@6,9:10=\"\n\",<WS>,L3:C2-L3:C3,chl=<H>,pl=<None>]"
- "[@7,10:11=\"&\",<MacroVarResolve>,L4:C0-L4:C1,chl=<D>,pl=0]"
- "[@8,11:15=\"mvar\",<MacroString>,L4:C1-L4:C5,chl=<D>,pl=<None>]"
- "[@9,15:16=\"\n\",<WS>,L4:C5-L4:C6,chl=<H>,pl=<None>]"
- "[@10,16:17=\"&\",<MacroVarResolve>,L5:C0-L5:C1,chl=<D>,pl=0]"
- "[@11,17:21=\"mvar\",<MacroString>,L5:C1-L5:C5,chl=<D>,pl=<None>]"
- "[@12,21:22=\".\",<MacroVarTerm>,L5:C5-L5:C6,chl=<D>,pl=<None>]"
- "[@13,22:23=\"\n\",<WS>,L5:C6-L5:C7,chl=<H>,pl=<None>]"
- "[@14,23:24=\"&\",<MacroVarResolve>,L6:C0-L6:C1,chl=<D>,pl=0]"
- "[@15,24:28=\"mvar\",<MacroString>,L6:C1-L6:C5,chl=<D>,pl=<None>]"
- "[@16,28:29=\"&\",<AMP>,L6:C5-L6:C6,chl=<D>,pl=<None>]"
- "[@17,29:30=\"\n\",<WS>,L6:C6-L6:C7,chl=<H>,pl=<None>]"
- "[@18,30:31=\"&\",<MacroVarResolve>,L7:C0-L7:C1,chl=<D>,pl=0]"
- "[@19,31:35=\"mvar\",<MacroString>,L7:C1-L7:C5,chl=<D>,pl=<None>]"
- "[@20,35:36=\".\",<MacroVarTerm>,L7:C5-L7:C6,chl=<D>,pl=<None>]"
- "[@21,36:37=\"&\",<AMP>,L7:C6-L7:C7,chl=<D>,pl=<None>]"
- "[@22,37:38=\"\n\",<WS>,L7:C7-L7:C8,chl=<H>,pl=<None>]"
- "[@23,38:39=\"&\",<MacroVarResolve>,L8:C0-L8:C1,chl=<D>,pl=0]"
- "[@24,39:43=\"mvar\",<MacroString>,L8:C1-L8:C5,chl=<D>,pl=<None>]"
- "[@25,43:44=\"&\",<AMP>,L8:C5-L8:C6,chl=<D>,pl=<None>]"
- "[@26,44:45=\".\",<DOT>,L8:C6-L8:C7,chl=<D>,pl=<None>]"
- "[@27,45:46=\"\n\",<WS>,L8:C7-L8:C8,chl=<H>,pl=<None>]"
- "[@28,46:48=\"&&\",<MacroVarResolve>,L9:C0-L9:C2,chl=<D>,pl=1]"
- "[@29,48:52=\"mvar\",<MacroString>,L9:C2-L9:C6,chl=<D>,pl=<None>]"
- "[@30,52:53=\"\n\",<WS>,L9:C6-L9:C7,chl=<H>,pl=<None>]"
- "[@31,53:55=\"&&\",<MacroVarResolve>,L10:C0-L10:C2,chl=<D>,pl=1]"
- "[@32,55:59=\"mvar\",<MacroString>,L10:C2-L10:C6,chl=<D>,pl=<None>]"
- "[@33,59:60=\".\",<MacroVarTerm>,L10:C6-L10:C7,chl=<D>,pl=<None>]"
- "[@34,60:61=\"\n\",<WS>,L10:C7-L10:C8,chl=<H>,pl=<None>]"
- "[@35,61:63=\"&&\",<MacroVarResolve>,L11:C0-L11:C2,chl=<D>,pl=1]"
- "[@36,63:67=\"mvar\",<MacroString>,L11:C2-L11:C6,chl=<D>,pl=<None>]"
- "[@37,67:68=\"&\",<AMP>,L11:C6-L11:C7,chl=<D>,pl=<None>]"
- "[@38,68:69=\"\n\",<WS>,L11:C7-L11:C8,chl=<H>,pl=<None>]"
- "[@39,69:71=\"&&\",<MacroVarResolve>,L12:C0-L12:C2,chl=<D>,pl=1]"
- "[@40,71:75=\"mvar\",<MacroString>,L12:C2-L12:C6,chl=<D>,pl=<None>]"
- "[@41,75:76=\".\",<MacroVarTerm>,L12:C6-L12:C7,chl=<D>,pl=<None>]"
- "[@42,76:77=\"&\",<AMP>,L12:C7-L12:C8,chl=<D>,pl=<None>]"
- "[@43,77:78=\"\n\",<WS>,L12:C8-L12:C9,chl=<H>,pl=<None>]"
- "[@44,78:80=\"&&\",<MacroVarResolve>,L13:C0-L13:C2,chl=<D>,pl=1]"
- "[@45,80:84=\"mvar\",<MacroString>,L13:C2-L13:C6,chl=<D>,pl=<None>]"
- "[@46,84:85=\"&\",<AMP>,L13:C6-L13:C7,chl=<D>,pl=<None>]"
- "[@47,85:86=\"9\",<IntegerLiteral>,L13:C7-L13:C8,chl=<D>,pl=9]"
- "[@48,86:87=\"\n\",<WS>,L13:C8-L13:C9,chl=<H>,pl=<None>]"
- "[@49,87:89=\"&&\",<MacroVarResolve>,L14:C0-L14:C2,chl=<D>,pl=1]"
- "[@50,89:93=\"mvar\",<MacroString>,L14:C2-L14:C6,chl=<D>,pl=<None>]"
- "[@51,93:94=\".\",<MacroVarTerm>,L14:C6-L14:C7,chl=<D>,pl=<None>]"
- "[@52,94:95=\"&\",<AMP>,L14:C7-L14:C8,chl=<D>,pl=<None>]"
- "[@53,95:96=\"9\",<IntegerLiteral>,L14:C8-L14:C9,chl=<D>,pl=9]"
- "[@54,96:97=\"\n\",<WS>,L14:C9-L14:C10,chl=<H>,pl=<None>]"
- "[@55,97:98=\"&\",<MacroVarResolve>,L15:C0-L15:C1,chl=<D>,pl=0]"
- "[@56,98:102=\"mvar\",<MacroString>,L15:C1-L15:C5,chl=<D>,pl=<None>]"
- "[@57,102:103=\"&\",<MacroVarResolve>,L15:C5-L15:C6,chl=<D>,pl=0]"
- "[@58,103:108=\"mvar2\",<MacroString>,L15:C6-L15:C11,chl=<D>,pl=<None>]"
- "[@59,108:109=\"\n\",<WS>,L15:C11-L15:C12,chl=<H>,pl=<None>]"
- "[@60,109:110=\"&\",<MacroVarResolve>,L16:C0-L16:C1,chl=<D>,pl=0]"
- "[@61,110:114=\"mvar\",<MacroString>,L16:C1-L16:C5,chl=<D>,pl=<None>]"
- "[@62,114:115=\"&\",<MacroVarResolve>,L16:C5-L16:C6,chl=<D>,pl=0]"
- "[@63,115:120=\"mvar2\",<MacroString>,L16:C6-L16:C11,chl=<D>,pl=<None>]"
- "[@64,120:121=\".\",<MacroVarTerm>,L16:C11-L16:C12,chl=<D>,pl=<None>]"
- "[@65,121:122=\"\n\",<WS>,L16:C12-L16:C13,chl=<H>,pl=<None>]"
- "[@66,122:123=\"&\",<MacroVarResolve>,L17:C0-L17:C1,chl=<D>,pl=0]"
- "[@67,123:127=\"mvar\",<MacroString>,L17:C1-L17:C5,chl=<D>,pl=<None>]"
- "[@68,127:128=\".\",<MacroVarTerm>,L17:C5-L17:C6,chl=<D>,pl=<None>]"
- "[@69,128:129=\"&\",<MacroVarResolve>,L17:C6-L17:C7,chl=<D>,pl=0]"
- "[@70,129:134=\"mvar2\",<MacroString>,L17:C7-L17:C12,chl=<D>,pl=<None>]"
- "[@71,134:135=\"\n\",<WS>,L17:C12-L17:C13,chl=<H>,pl=<None>]"
- "[@72,135:136=\"&\",<MacroVarResolve>,L18:C0-L18:C1,chl=<D>,pl=0]"
- "[@73,136:140=\"mvar\",<MacroString>,L18:C1-L18:C5,chl=<D>,pl=<None>]"
- "[@74,140:141=\".\",<MacroVarTerm>,L18:C5-L18:C6,chl=<D>,pl=<None>]"
- "[@75,141:142=\"&\",<MacroVarResolve>,L18:C6-L18:C7,chl=<D>,pl=0]"
- "[@76,142:147=\"mvar2\",<MacroString>,L18:C7-L18:C12,chl=<D>,pl=<None>]"
- "[@77,147:148=\".\",<MacroVarTerm>,L18:C12-L18:C13,chl=<D>,pl=<None>]"
- "[@78,148:149=\"\n\",<WS>,L18:C13-L18:C14,chl=<H>,pl=<None>]"
- "[@79,149:151=\"&&\",<MacroVarResolve>,L19:C0-L19:C2,chl=<D>,pl=1]"
- "[@80,151:155=\"mvar\",<MacroString>,L19:C2-L19:C6,chl=<D>,pl=<None>]"
- "[@81,155:156=\"&\",<MacroVarResolve>,L19:C6-L19:C7,chl=<D>,pl=0]"
- "[@82,156:161=\"mvar2\",<MacroString>,L19:C7-L19:C12,chl=<D>,pl=<None>]"
- "[@83,161:162=\"\n\",<WS>,L19:C12-L19:C13,chl=<H>,pl=<None>]"
- "[@84,162:164=\"&&\",<MacroVarResolve>,L20:C0-L20:C2,chl=<D>,pl=1]"
- "[@85,164:168=\"mvar\",<MacroString>,L20:C2-L20:C6,chl=<D>,pl=<None>]"
- "[@86,168:169=\"&\",<MacroVarResolve>,L20:C6-L20:C7,chl=<D>,pl=0]"
- "[@87,169:174=\"mvar2\",<MacroString>,L20:C7-L20:C12,chl=<D>,pl=<None>]"
- "[@88,174:175=\".\",<MacroVarTerm>,L20:C12-L20:C13,chl=<D>,pl=<None>]"
- "[@89,175:176=\"\n\",<WS>,L20:C13-L20:C14,chl=<H>,pl=<None>]"
- "[@90,176:178=\"&&\",<MacroVarResolve>,L21:C0-L21:C2,chl=<D>,pl=1]"
- "[@91,178:182=\"mvar\",<MacroString>,L21:C2-L21:C6,chl=<D>,pl=<None>]"
- "[@92,182:183=\".\",<MacroVarTerm>,L21:C6-L21:C7,chl=<D>,pl=<None>]"
- "[@93,183:184=\"&\",<MacroVarResolve>,L21:C7-L21:C8,chl=<D>,pl=0]"
- "[@94,184:189=\"mvar2\",<MacroString>,L21:C8-L21:C13,chl=<D>,pl=<None>]"
- "[@95,189:190=\"\n\",<WS>,L21:C13-L21:C14,chl=<H>,pl=<None>]"
- "[@96,190:192=\"&&\",<MacroVarResolve>,L22:C0-L22:C2,chl=<D>,pl=1]"
- "[@97,192:196=\"mvar\",<MacroString>,L22:C2-L22:C6,chl=<D>,pl=<None>]"
- "[@98,196:197=\".\",<MacroVarTerm>,L22:C6-L22:C7,chl=<D>,pl=<None>]"
- "[@99,197:198=\"&\",<MacroVarResolve>,L22:C7-L22:C8,chl=<D>,pl=0]"
- "[@100,198:203=\"mvar2\",<MacroString>,L22:C8-L22:C13,chl=<D>,pl=<None>]"
- "[@101,203:204=\".\",<MacroVarTerm>,L22:C13-L22:C14,chl=<D>,pl=<None>]"
- "[@102,204:205=\"\n\",<WS>,L22:C14-L22:C15,chl=<H>,pl=<None>]"
- "[@103,205:207=\"&&\",<MacroVarResolve>,L23:C0-L23:C2,chl=<D>,pl=1]"
- "[@104,207:208=\"&\",<MacroVarResolve>,L23:C2-L23:C3,chl=<D>,pl=0]"
- "[@105,208:212=\"mvar\",<MacroString>,L23:C3-L23:C7,chl=<D>,pl=<None>]"
- "[@106,212:213=\"&\",<MacroVarResolve>,L23:C7-L23:C8,chl=<D>,pl=0]"
- "[@107,213:218=\"mvar2\",<MacroString>,L23:C8-L23:C13,chl=<D>,pl=<None>]"
- "[@108,218:219=\"\n\",<WS>,L23:C13-L23:C14,chl=<H>,pl=<None>]"
- "[@109,219:221=\"&&\",<MacroVarResolve>,L24:C0-L24:C2,chl=<D>,pl=1]"
- "[@110,221:222=\"&\",<MacroVarResolve>,L24:C2-L24:C3,chl=<D>,pl=0]"
- "[@111,222:226=\"mvar\",<MacroString>,L24:C3-L24:C7,chl=<D>,pl=<None>]"
- "[@112,226:227=\"&\",<MacroVarResolve>,L24:C7-L24:C8,chl=<D>,pl=0]"
- "[@113,227:232=\"mvar2\",<MacroString>,L24:C8-L24:C13,chl=<D>,pl=<None>]"
- "[@114,232:233=\".\",<MacroVarTerm>,L24:C13-L24:C14,chl=<D>,pl=<None>]"
- "[@115,233:234=\"\n\",<WS>,L24:C14-L24:C15,chl=<H>,pl=<None>]"
- "[@116,234:236=\"&&\",<MacroVarResolve>,L25:C0-L25:C2,chl=<D>,pl=1]"
- "[@117,236:237=\"&\",<MacroVarResolve>,L25:C2-L25:C3,chl=<D>,pl=0]"
- "[@118,237:241=\"mvar\",<MacroString>,L25:C3-L25:C7,chl=<D>,pl=<None>]"
- "[@119,241:242=\".\",<MacroVarTerm>,L25:C7-L25:C8,chl=<D>,pl=<None>]"
- "[@120,242:243=\"&\",<MacroVarResolve>,L25:C8-L25:C9,chl=<D>,pl=0]"
- "[@121,243:248=\"mvar2\",<MacroString>,L25:C9-L25:C14,chl=<D>,pl=<None>]"
- "[@122,248:249=\"\n\",<WS>,L25:C14-L25:C15,chl=<H>,pl=<None>]"
- "[@123,249:251=\"&&\",<MacroVarResolve>,L26:C0-L26:C2,chl=<D>,pl=1]"
- "[@124,251:252=\"&\",<MacroVarResolve>,L26:C2-L26:C3,chl=<D>,pl=0]"
- "[@125,252:256=\"mvar\",<MacroString>,L26:C3-L26:C7,chl=<D>,pl=<None>]"
- "[@126,256:257=\".\",<MacroVarTerm>,L26:C7-L26:C8,chl=<D>,pl=<None>]"
- "[@127,257:258=\"&\",<MacroVarResolve>,L26:C8-L26:C9,chl=<D>,pl=0]"
- "[@128,258:263=\"mvar2\",<MacroString>,L26:C9-L26:C14,chl=<D>,pl=<None>]"
- "[@129,263:264=\".\",<MacroVarTerm>,L26:C14-L26:C15,chl=<D>,pl=<None>]"
- "[@130,264:265=\"\n\",<WS>,L26:C15-L26:C16,chl=<H>,pl=<None>]"
- "[@131,265:266=\"&\",<MacroVarResolve>,L27:C0-L27:C1,chl=<D>,pl=0]"
- "[@132,266:267=\"m\",<MacroString>,L27:C1-L27:C2,chl=<D>,pl=<None>]"
- "[@133,267:268=\".\",<MacroVarTerm>,L27:C2-L27:C3,chl=<D>,pl=<None>]"
- "[@134,268:269=\"_\",<Identifier>,L27:C3-L27:C4,chl=<D>,pl=<None>]"
- "[@135,269:270=\"\n\",<WS>,L27:C4-L27:C5,chl=<H>,pl=<None>]"
- "[@136,270:271=\"&\",<MacroVarResolve>,L28:C0-L28:C1,chl=<D>,pl=0]"
- "[@137,271:272=\"m\",<MacroString>,L28:C1-L28:C2,chl=<D>,pl=<None>]"
- "[@138,272:273=\".\",<MacroVarTerm>,L28:C2-L28:C3,chl=<D>,pl=<None>]"
- "[@139,273:274=\".\",<DOT>,L28:C3-L28:C4,chl=<D>,pl=<None>]"
- "[@140,274:275=\"_\",<Identifier>,L28:C4-L28:C5,chl=<D>,pl=<None>]"
- "[@141,275:276=\"\n\",<WS>,L28:C5-L28:C6,chl=<H>,pl=<None>]"
- "[@142,276:278=\"&&\",<MacroVarResolve>,L29:C0-L29:C2,chl=<D>,pl=1]"
- "[@143,278:279=\"p\",<MacroString>,L29:C2-L29:C3,chl=<D>,pl=<None>]"
- "[@144,279:280=\"&\",<MacroVarResolve>,L29:C3-L29:C4,chl=<D>,pl=0]"
- "[@145,280:281=\"m\",<MacroString>,L29:C4-L29:C5,chl=<D>,pl=<None>]"
- "[@146,281:282=\".\",<MacroVarTerm>,L29:C5-L29:C6,chl=<D>,pl=<None>]"
- "[@147,282:283=\"_\",<MacroString>,L29:C6-L29:C7,chl=<D>,pl=<None>]"
- "[@148,283:284=\"\n\",<WS>,L29:C7-L29:C8,chl=<H>,pl=<None>]"
- "[@149,284:286=\"&&\",<MacroVarResolve>,L30:C0-L30:C2,chl=<D>,pl=1]"
- "[@150,286:287=\"p\",<MacroString>,L30:C2-L30:C3,chl=<D>,pl=<None>]"
- "[@151,287:288=\"&\",<MacroVarResolve>,L30:C3-L30:C4,chl=<D>,pl=0]"
- "[@152,288:289=\"m\",<MacroString>,L30:C4-L30:C5,chl=<D>,pl=<None>]"
- "[@153,289:290=\".\",<MacroVarTerm>,L30:C5-L30:C6,chl=<D>,pl=<None>]"
- "[@154,290:291=\".\",<MacroVarTerm>,L30:C6-L30:C7,chl=<D>,pl=<None>]"
- "[@155,291:292=\"_\",<Identifier>,L30:C7-L30:C8,chl=<D>,pl=<None>]"
- "[@156,292:293=\"\n\",<WS>,L30:C8-L30:C9,chl=<H>,pl=<None>]"
- "[@157,293:295=\"&&\",<MacroVarResolve>,L31:C0-L31:C2,chl=<D>,pl=1]"
- "[@158,295:296=\"p\",<MacroString>,L31:C2-L31:C3,chl=<D>,pl=<None>]"
- "[@159,296:298=\"&&\",<MacroVarResolve>,L31:C3-L31:C5,chl=<D>,pl=1]"
- "[@160,298:299=\"m\",<MacroString>,L31:C5-L31:C6,chl=<D>,pl=<None>]"
- "[@161,299:300=\".\",<MacroVarTerm>,L31:C6-L31:C7,chl=<D>,pl=<None>]"
- "[@162,300:301=\"_\",<Identifier>,L31:C7-L31:C8,chl=<D>,pl=<None>]"
- "[@163,301:302=\"\n\",<WS>,L31:C8-L31:C9,chl=<H>,pl=<None>]"
- "[@164,302:304=\"&&\",<MacroVarResolve>,L32:C0-L32:C2,chl=<D>,pl=1]"
- "[@165,304:305=\"p\",<MacroString>,L32:C2-L32:C3,chl=<D>,pl=<None>]"
- "[@166,305:307=\"&&\",<MacroVarResolve>,L32:C3-L32:C5,chl=<D>,pl=1]"
- "[@167,307:308=\"m\",<MacroString>,L32:C5-L32:C6,chl=<D>,pl=<None>]"
- "[@168,308:309=\".\",<MacroVarTerm>,L32:C6-L32:C7,chl=<D>,pl=<None>]"
- "[@169,309:310=\".\",<DOT>,L32:C7-L32:C8,chl=<D>,pl=<None>]"
- "[@170,310:311=\"_\",<Identifier>,L32:C8-L32:C9,chl=<D>,pl=<None>]"
- "[@171,311:312=\"\n\",<WS>,L32:C9-L32:C10,chl=<H>,pl=<None>]"
- "[@172,312:316=\"&&&&\",<MacroVarResolve>,L33:C0-L33:C4,chl=<D>,pl=2]"
- "[@173,316:317=\"p\",<MacroString>,L33:C4-L33:C5,chl=<D>,pl=<None>]"
- "[@174,317:319=\"&&\",<MacroVarResolve>,L33:C5-L33:C7,chl=<D>,pl=1]"
- "[@175,319:320=\"p\",<MacroString>,L33:C7-L33:C8,chl=<D>,pl=<None>]"
- "[@176,320:321=\"&\",<MacroVarResolve>,L33:C8-L33:C9,chl=<D>,pl=0]"
- "[@177,321:322=\"m\",<MacroString>,L33:C9-L33:C10,chl=<D>,pl=<None>]"
- "[@178,322:323=\".\",<MacroVarTerm>,L33:C10-L33:C11,chl=<D>,pl=<None>]"
- "[@179,323:324=\".\",<MacroVarTerm>,L33:C11-L33:C12,chl=<D>,pl=<None>]"
- "[@180,324:325=\".\",<MacroVarTerm>,L33:C12-L33:C13,chl=<D>,pl=<None>]"
- "[@181,325:326=\"_\",<Identifier>,L33:C13-L33:C14,chl=<D>,pl=<None>]"
- "[@182,326:327=\"\n\",<WS>,L33:C14-L33:C15,chl=<H>,pl=<None>]"
- "[@183,327:331=\"&&&&\",<MacroVarResolve>,L34:C0-L34:C4,chl=<D>,pl=2]"
- "[@184,331:332=\"p\",<MacroString>,L34:C4-L34:C5,chl=<D>,pl=<None>]"
- "[@185,332:333=\".\",<MacroVarTerm>,L34:C5-L34:C6,chl=<D>,pl=<None>]"
- "[@186,333:335=\"&&\",<MacroVarResolve>,L34:C6-L34:C8,chl=<D>,pl=1]"
- "[@187,335:336=\"p\",<MacroString>,L34:C8-L34:C9,chl=<D>,pl=<None>]"
- "[@188,336:337=\"&\",<MacroVarResolve>,L34:C9-L34:C10,chl=<D>,pl=0]"
- "[@189,337:338=\"m\",<MacroString>,L34:C10-L34:C11,chl=<D>,pl=<None>]"
- "[@190,338:339=\".\",<MacroVarTerm>,L34:C11-L34:C12,chl=<D>,pl=<None>]"
- "[@191,339:340=\".\",<MacroVarTerm>,L34:C12-L34:C13,chl=<D>,pl=<None>]"
- "[@192,340:341=\"_\",<Identifier>,L34:C13-L34:C14,chl=<D>,pl=<None>]"
- "[@193,341:342=\"\n\",<WS>,L34:C14-L34:C15,chl=<H>,pl=<None>]"
- "[@194,342:346=\"&&&&\",<MacroVarResolve>,L35:C0-L35:C4,chl=<D>,pl=2]"
- "[@195,346:347=\"p\",<MacroString>,L35:C4-L35:C5,chl=<D>,pl=<None>]"
- "[@196,347:349=\"&&\",<MacroVarResolve>,L35:C5-L35:C7,chl=<D>,pl=1]"
- "[@197,349:350=\"p\",<MacroString>,L35:C7-L35:C8,chl=<D>,pl=<None>]"
- "[@198,350:351=\".\",<MacroVarTerm>,L35:C8-L35:C9,chl=<D>,pl=<None>]"
- "[@199,351:352=\"&\",<MacroVarResolve>,L35:C9-L35:C10,chl=<D>,pl=0]"
- "[@200,352:353=\"m\",<MacroString>,L35:C10-L35:C11,chl=<D>,pl=<None>]"
- "[@201,353:354=\".\",<MacroVarTerm>,L35:C11-L35:C12,chl=<D>,pl=<None>]"
- "[@202,354:355=\".\",<MacroVarTerm>,L35:C12-L35:C13,chl=<D>,pl=<None>]"
- "[@203,355:356=\"_\",<Identifier>,L35:C13-L35:C14,chl=<D>,pl=<None>]"
- "[@204,356:357=\"\n\",<WS>,L35:C14-L35:C15,chl=<H>,pl=<None>]"
- "[@205,357:358=\"&\",<MacroVarResolve>,L36:C0-L36:C1,chl=<D>,pl=0]"
- "[@206,358:359=\"i\",<MacroString>,L36:C1-L36:C2,chl=<D>,pl=<None>]"
- "[@207,359:360=\"\n\",<WS>,L36:C2-L36:C3,chl=<H>,pl=<None>]"
- "[@208,360:362=\"&&\",<MacroVarResolve>,L37:C0-L37:C2,chl=<D>,pl=1]"
- "[@209,362:363=\"i\",<MacroString>,L37:C2-L37:C3,chl=<D>,pl=<None>]"
- "[@210,363:364=\"\n\",<WS>,L37:C3-L37:C4,chl=<H>,pl=<None>]"
- "[@211,364:366=\"&&\",<MacroVarResolve>,L38:C0-L38:C2,chl=<D>,pl=1]"
- "[@212,366:367=\"&\",<MacroVarResolve>,L38:C2-L38:C3,chl=<D>,pl=0]"
- "[@213,367:368=\"i\",<MacroString>,L38:C3-L38:C4,chl=<D>,pl=<None>]"
- "[@214,368:369=\"\n\",<WS>,L38:C4-L38:C5,chl=<H>,pl=<None>]"
- "[@215,369:373=\"&&&&\",<MacroVarResolve>,L39:C0-L39:C4,chl=<D>,pl=2]"
- "[@216,373:375=\"&&\",<MacroVarResolve>,L39:C4-L39:C6,chl=<D>,pl=1]"
- "[@217,375:376=\"&\",<MacroVarResolve>,L39:C6-L39:C7,chl=<D>,pl=0]"
- "[@218,376:377=\"i\",<MacroString>,L39:C7-L39:C8,chl=<D>,pl=<None>]"
- "[@219,377:378=\"\n\",<WS>,L39:C8-L39:C9,chl=<H>,pl=<None>]"
- "[@220,378:382=\"&&&&\",<MacroVarResolve>,L40:C0-L40:C4,chl=<D>,pl=2]"
- "[@221,382:383=\"p\",<MacroString>,L40:C4-L40:C5,chl=<D>,pl=<None>]"
- "[@222,383:385=\"&&\",<MacroVarResolve>,L40:C5-L40:C7,chl=<D>,pl=1]"
- "[@223,385:386=\"p\",<MacroString>,L40:C7-L40:C8,chl=<D>,pl=<None>]"
- "[@224,386:387=\"&\",<MacroVarResolve>,L40:C8-L40:C9,chl=<D>,pl=0]"
- "[@225,387:388=\"m\",<MacroString>,L40:C9-L40:C10,chl=<D>,pl=<None>]"
- "[@226,388:389=\".\",<MacroVarTerm>,L40:C10-L40:C11,chl=<D>,pl=<None>]"
- "[@227,389:390=\".\",<MacroVarTerm>,L40:C11-L40:C12,chl=<D>,pl=<None>]"
- "[@228,390:391=\"_\",<MacroString>,L40:C12-L40:C13,chl=<D>,pl=<None>]"
- "[@229,391:392=\"\n\",<WS>,L40:C13-L40:C14,chl=<H>,pl=<None>]"
- "[@230,392:394=\"&&\",<MacroVarResolve>,L41:C0-L41:C2,chl=<D>,pl=1]"
- "[@231,394:396=\"m_\",<MacroString>,L41:C2-L41:C4,chl=<D>,pl=<None>]"
- "[@232,396:398=\"&&\",<MacroVarResolve>,L41:C4-L41:C6,chl=<D>,pl=1]"
- "[@233,398:399=\"p\",<MacroString>,L41:C6-L41:C7,chl=<D>,pl=<None>]"
- "[@234,399:400=\"&\",<MacroVarResolve>,L41:C7-L41:C8,chl=<D>,pl=0]"
- "[@235,400:401=\"m\",<MacroString>,L41:C8-L41:C9,chl=<D>,pl=<None>]"
- "[@236,401:402=\".\",<MacroVarTerm>,L41:C9-L41:C10,chl=<D>,pl=<None>]"
- "[@237,402:403=\".\",<MacroVarTerm>,L41:C10-L41:C11,chl=<D>,pl=<None>]"
- "[@238,403:404=\"_\",<Identifier>,L41:C11-L41:C12,chl=<D>,pl=<None>]"
- "[@239,404:405=\"\n\",<WS>,L41:C12-L41:C13,chl=<H>,pl=<None>]"
- "[@240,405:405=\"<no text>\",<EOF>,L42:C0-L42:C0,chl=<D>,pl=<None>]"
